Siwx91x dma driver fixes

smalae opened the pull request:
The following fixes are done in the DMA driver:

1. Introduced a new variable in the `dma_siwx91x_channel_info`
structure to provide a clean way to differentiate transfer
directions. This enhancement is utilized to trigger software
requests specifically for memory-to-memory transfers
2. Addressed an issue where regular/non-scatter-gather DMA
transfers were not explicitly using the primary DMA descriptor
structure. This ensures a smooth regular DMA transfer after
any scatter-gather transfer.
3. Addressed an issue where alignment of dma desc varaible
of dma0 is corrected to 1024 instead of 512.

smalae committed:
drivers: dma: siwx91x: distinguishing mem to mem transfers

Introduced a new variable in the `dma_siwx91x_channel_info`
structure to provide a clean way to differentiate transfer
directions. This enhancement is utilized to trigger software
requests specifically for memory-to-memory transfers

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

jhedberg commented on drivers/dma/dma_silabs_siwx91x.c:
```diff
@@ -37,6 +37,7 @@ struct dma_siwx91x_channel_info {
 	dma_callback_t dma_callback;        /* User callback */
 	void *cb_data;                      /* User callback data */
 	RSI_UDMA_DESC_T *sg_desc_addr_info; /* Scatter-Gather table start address */
+	uint8_t transfer_direction;         /* mem<->mem ot per<->mem */
```
I think it's worth to consider giving a named type for the enum and just use that everywhere instead of the mix of `uint8_t` and `int`. Btw, you don't gain anything with using a smaller type in the above struct due to implicit padding, i.e. the struct size would be exactly the same if you used `int`. Even if you did need a smaller type, you can also get that with an enum by adding `__packed` to the definition. I'd also consider adding a third possible value to make the code look more intuitive (e.g. you can return an actual enum value instead of `-EINVAL`):
```c
enum transfer_dir {
	TRANSFER_MEM_TO_MEM,
	TRANSFER_TO_OR_FROM_PER,
	TRANSFER_DIR_INVALID = -1,
};
```

Perhaps out of scope for this PR, but if you want to make these symbol names more compact, I believe a common convention is to shorten "transfer" to "xfer'.

smalae replied:
I've updated it according to the suggestion.

jerome-pouiller commented on drivers/dma/dma_silabs_siwx91x.c:
```diff
@@ -464,8 +478,7 @@ static int siwx91x_dma_start(const struct device *dev, uint32_t channel)
 	/* Check if the transfer type is memory-memory */
-	if (udma_table[channel].vsUDMAChaConfigData1.srcInc != UDMA_SRC_INC_NONE &&
-	    udma_table[channel].vsUDMAChaConfigData1.dstInc != UDMA_DST_INC_NONE) {
+	if (data->zephyr_channel_info[channel].xfer_direction == TRANSFER_MEM_TO_MEM) {
```
Just curiosity, in which case the original condition didn't work?

smalae replied:
In case of transfers using alternate descriptors, this condition is true for all the transfers and triggers the SW request even for peripheral transfers.

jhedberg commented on drivers/dma/dma_silabs_siwx91x.c:
```diff
@@ -206,13 +208,13 @@ static int siwx91x_sg_chan_config(const struct device *dev, RSI_UDMA_HANDLE_T ud
 	RSI_UDMA_DESC_T *sg_desc_base_addr = NULL;
 	uint8_t transfer_type;
-	int ret;
+	int xfer_dir;
```
```c
enum xfer_dir dir;
```

smalae replied:
Updated

jhedberg commented on drivers/dma/dma_silabs_siwx91x.c:
```diff
@@ -206,13 +208,13 @@ static int siwx91x_sg_chan_config(const struct device *dev, RSI_UDMA_HANDLE_T ud
-	ret = siwx91x_transfer_direction(config->channel_direction);
-	if (ret < 0) {
+	xfer_dir = siwx91x_transfer_direction(config->channel_direction);
+	if (xfer_dir < 0) {
```
```c
if (dir == TRANSFER_DIR_INVALID) {
```

smalae replied:
Done

jhedberg commented on drivers/dma/dma_silabs_siwx91x.c:
```diff
@@ -206,13 +208,13 @@ static int siwx91x_sg_chan_config(const struct device *dev, RSI_UDMA_HANDLE_T ud
 		return -EINVAL;
 	}
-	transfer_type = ret ? UDMA_MODE_PER_SCATTER_GATHER : UDMA_MODE_MEM_SCATTER_GATHER;
+	transfer_type = xfer_dir ? UDMA_MODE_PER_SCATTER_GATHER : UDMA_MODE_MEM_SCATTER_GATHER;
```
```c
if (dir == TRANSFER_TO_OR_FROM_PER) {
	transfer_type = UDMA_MODE_PER_SCATTER_GATHER;
} else {
	transfer_type = UDMA_MODE_MEM_SCATTER_GATHER;
}
```

smalae replied:
Updated

smalae committed:
drivers: dma: siwx91x: Bug fix for regular DMA transfers

Addressed an issue where regular/non-scatter-gather DMA
transfers were not explicitly using the primary DMA descriptor
structure. This ensures a smooth regular DMA transfer after
any scatter gather transfer.

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

jerome-pouiller commented on drivers/dma/dma_silabs_siwx91x.c:
```diff
@@ -28,15 +28,17 @@
 LOG_MODULE_REGISTER(si91x_dma, CONFIG_DMA_LOG_LEVEL);
 
-enum {
+enum xfer_dir_t {
```
Zephyr tends to apply the Coding Style of the Linux kernel. So, `_t` is only used for typedefs. In addition, we try to avoid typedefs and use plain `enum xx` or `struct yy` instead.

smalae replied:
Updated

smalae committed:
drivers: dma: siwx91x: SRAM desc alignment bug fix

Addressed an issue where alignment of dma desc varaible
of dma0 is corrected to 1024 instead of 512.

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae commented:
@jerome-pouiller @jhedberg . I have added one more commit to this PR, which corrects the alignment of sram_desc_addr from 512 to 1024. This issue was identified during SPI testing with the latest DMA changes introduced by the memory_partition PR. Kindly review.

jhedberg commented on drivers/dma/dma_silabs_siwx91x.c:
```diff
@@ -68,7 +70,7 @@ static int siwx91x_transfer_direction(uint32_t dir)
 	}
 
-	return -EINVAL;
+	return TRANSFER_DIR_INVALID;
```
You still need to update the return value type of `siwx91x_transfer_direction()` from `int` to `enum dma_xfer_dir`

jhedberg commented on drivers/dma/dma_silabs_siwx91x.c:
```diff
@@ -28,15 +28,17 @@
 	dma_callback_t dma_callback;        /* User callback */
 	void *cb_data;                      /* User callback data */
 	RSI_UDMA_DESC_T *sg_desc_addr_info; /* Scatter-Gather table start address */
+	int xfer_direction;                 /* mem<->mem ot per<->mem */
```
One more `int` -> `enum dma_xfer_dir` here
